calamari_ocr/ocr/datasets/pagexml_dataset/dataset.py

```diff
@@ -254,12 +255,73 @@ def cutout(pageimg, coordstring, scale=1, rect=False):
         box[rr - offset[0], cc - offset[1]] = pageimg[rr, cc]
         return box
 
+    @staticmethod
+    def get_words(prediction, sample):
+        def remove_leading_spaces(_positions):
+            return list(itertools.dropwhile(lambda p: p[0][0] == " ", _positions))
+
+        def remove_trailing_spaces(_positions):
+            return list(reversed(remove_leading_spaces(reversed(_positions))))
+
+        x_coords, y_coords = map(list, zip(*[coord.split(",") for coord in sample['coords'].split()]))
+        x, y = [int(x) for x in x_coords], [int(y) for y in y_coords]
+        min_x, max_x, min_y, max_y = min(x), max(x), min(y), max(y)
+
+        positions = [(pos.chars[0].char, pos.global_start + min_x, pos.global_end + min_x) for pos in
+                     prediction.positions]
+        positions = remove_leading_spaces(positions)
+        positions = remove_trailing_spaces(positions)
+
+        words = [{"char": positions[0][0], "min_x": positions[0][1], "max_x": positions[0][2],
+                  "min_y": min_y, "max_y": max_y}]
+        new_word = False
+
+        for pos, entry in enumerate(positions[1:]):
+            if entry[0] == " ":
+                words.append({"char": entry[0], "min_x": entry[1], "max_x": entry[2], "min_y": min_y, "max_y": max_y})
+                new_word = True
+                continue
+            if new_word:
+                words.append({"char": entry[0], "min_x": entry[1], "max_x": entry[2], "min_y": min_y, "max_y": max_y})
+                new_word = False
+            else:
+                words[-1]["char"] += entry[0]
+                words[-1]["max_x"] = entry[2]
+
+        return words
+
+    def store_words(self, prediction, sample):
+        ns = sample['ns']
+        line = sample['xml_element']
+
+        line_id = line.attrib["id"]
+        textequivelem = line.find("./ns:TextEquiv", namespaces=ns)
+
+        words = self.get_words(prediction, sample)
+
+        if textequivelem is not None:
+            for index, word in enumerate(words, 1):
+                word_elem = etree.Element("Word", id=f"{line_id}_w{str(index).zfill(3)}")
```

Comment:
Are the <word>-Elements in some cases needed for search with .find or .xpath later (eg. second run, replacement before writing to disk?) Then we would need something like` etree.SubElement("{%s}Word" % ns["ns"], nsmap=ns)` here. Otherwise elements without explicit namespace declaration are found only after writing to disk and re-parsing the file.

Reply:
I didn't think about such a case, but there probably aren't any downsides to declaring the namespace just in case it might be useful later on, are there?

Reply:
I hope so. Maybe we should test it with explicit declaration and run some tests to be sure there are not unwanted traces of redundant namespaces left in the XML files.

Reply:
Ran some tests with explicit namespace declaration and it didn't cause any problems in the output XML files.